### PR TITLE
data/aws: Bump default instance type from t2.medium to t3.medium

### PR DESCRIPTION
**Pull request by wking**
The T3 instances [were launched in 2018-08-21][1].  [The documents stats][3] are fairly similar to [the T2 analogs][2].  The only obvious difference there is that the T3 instances are 10% cheaper.  However, from [here][1]:

> If the instance runs at higher CPU utilization for a prolonged period, there will be an additional charge of $0.05 per vCPU-hour.

and from [here][2]:

> T3 instances start in Unlimited mode by default... If the instance needs to run at higher CPU utilization for a prolonged period, it can do so for an additional charge of $0.05 per vCPU-hour.  In  Standard mode, a T3 instance can burst until it uses up all its earned credits.

The difference vs. T2 seems to be that with T2, Unlimited mode was opt-in while with T3 it is opt-out.  I've set [a credit specification in Terraform][5] to opt us out, but there doesn't seem to be a setting in the cluster-API structures for this.

[1]: https://aws.amazon.com/blogs/aws/new-t3-instances-burstable-cost-effective-performance/
[2]: https://aws.amazon.com/ec2/instance-types/t2/
[3]: https://aws.amazon.com/ec2/instance-types/t3/
[5]: https://www.terraform.io/docs/providers/aws/r/instance.html#credit-specification

### CODE REVIEW

File: data/data/aws/bootstrap/main.tf [hunk 1]
```diff
@@ -128,6 +128,10 @@ resource "aws_instance" "bootstrap" {
   }
 
   volume_tags = "${var.tags}"
+
+  credit_specification {
+    cpu_credits = "standard"
```
Comment (cuppett) [inline data/data/aws/bootstrap/main.tf]: I would recommend unlimited here. I don't know we're optimizing purely for cost. Developer productivity and cluster health are critically important right now. If we exhaust credits during startup of the node and operators, we don't want it to take excessively long to get a stable cluster due to throttling. Also, if we're constantly out of credits for clusters over time, the c5.large is likely a better selection. This assumes this is the correct entry form-factor (2 CPUs and 4 GB memory); we should optimize for performance density of the node type selected.

**Comment by cuppett** [comment]
> I don't think we want to move towards more reliability on defaults in terraform. Also this causes the master machine definitions to drift from the machines that actually exist.

Should the master machine definitions be changed as well? We must assume our default node types will change over time (even after the initial 4.0 release).

**Comment by wking** [comment]
> Should the master machine definitions be changed as well?

This same default [feeds the master definitions][1].

[1]: https://github.com/openshift/installer/blob/v0.4.0/pkg/asset/machines/master.go#L63

File: data/data/aws/bootstrap/main.tf (hunk 1 above)
Comment (cuppett) [reply data/data/aws/bootstrap/main.tf]: For our masters in OSO starter, we run 3x r4.4xlarge nodes. Customizing these node types will be a common need for the target size of the cluster. Ours stay around 40% CPU utilized. We can send you other statistics/metrics if you are interested.

Comment (wking) [reply data/data/aws/bootstrap/main.tf]: > Customizing these node types will be a common need for the target size of the cluster.

Users can customize already, e.g. by setting [type][1] in their install-config during [a multi-stage install][2].  What's currently missing are knobs in the cluster-API types for detailed tuning like this CPU-credit handling.

[1]: https://github.com/openshift/installer/blob/v0.4.0/pkg/types/aws/machinepool.go#L12-L14
[2]: https://github.com/openshift/installer/blob/v0.4.0/docs/user/overview.md#multiple-invocations

**Comment by smarterclayton** [comment]
let's set some requirements:

1. the defaults for openshift should work out of the box for a reasonable period of time for most users (including smaller users than we traditionally encouraged)
2. in the future we expect to have master rotation, at which point vertical resize is free, so we shouldn't overdesign here
3. we want to test what we give to our users, so there is an argument for e2e being the high end of what we support for those users (we are confident that if e2e is not flaky our users will be able to use the ootb cluster)

3-15 nodes is a good small-medium cluster, 100 namespaces.  e2e pushes harder than that on API traffic, but runs on 3 nodes densely.

File: data/data/aws/bootstrap/main.tf (hunk 1 above)
Comment (wking) [reply data/data/aws/bootstrap/main.tf]: Ok, I've rebased onto master and dropped the `cpu_credits` block with 7d236d0 -> 3bdf83b.  I've tried to summarize the motivation for leaving Unlimited mode on towards the end of the 3bdf83b commit message.

**Comment by wking** [comment]
> I don't think we want to move towards more reliability on defaults in terraform.

With 7d236d0 -> 3bdf83b, I've dropped some more Terraform defaults to make sure we aren't relying on them.  I've also [dropped the `cpu_credits` block][1].  That should address your concerns on those points.

We'll see how 3bdf83b plays in CI, since we're not sure the initial CPU credits are sufficient to get us through the e2e tests without paying too much for the extended burst.

[1]: https://github.com/openshift/installer/pull/724#discussion_r236801498

**Comment by wking** [comment]
[e2e-aws][1]:

```
level=error msg="Error: Required variable not set: tectonic_aws_master_ec2_type\n\n\n\nError: Required variable not set: tectonic_aws_worker_ec2_type"
```

I'm patching to get that passed through...

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/724/pull-ci-openshift-installer-master-e2e-aws/1644/build-log.txt

**Comment by wking** [comment]
Actually, we probably don't need `tectonic_aws_worker_ec2_type` anymore...

**Comment by wking** [comment]
> Actually, we probably don't need `tectonic_aws_worker_ec2_type` anymore...

I've filed #742 for this.

**Comment by abhinavdahiya** [comment]
@wking can we take https://github.com/openshift/installer/pull/724/commits/2b76d620e2322a041f04cc9201956c6f98fc9744 out from this PR. lets just bump t2 -> t3 here (nothing worse than we have right now in terms of coordinating values b/w terraform and cluster-api). 

https://github.com/openshift/installer/pull/724/commits/2b76d620e2322a041f04cc9201956c6f98fc9744 in separate PR.

**Comment by wking** [comment]
I've pushed 85e86cf -> 30521b4, dropping the `TFVars` refactor at @abhinavdahiya's [request][1].  I've left the commits from #742 underneath, but without the `TFVars` refactor to motivate them, I could also cherry-pick 30521b4 onto master by itself if folks wanted it that way.  I'll push 85e86cf to a separate PR once CI greens up so we can fix the dependency disconnect.

[1]: https://github.com/openshift/installer/pull/724#issuecomment-442959144

**Comment by wking** [comment]
[e2e-aws][1]:

```
level=error msg="Error: Error applying plan:\n\n1 error(s) occurred:\n\n* aws_route53_zone.tectonic_int: 1 error(s) occurred:\n\n* aws_route53_zone.tectonic_int: timeout while waiting for state to become 'INSYNC' (timeout: 15m0s)\n\nTerraform does not automatically rollback in the face of errors.\nInstead, your Terraform state file has been partially updated with\nany resources that successfully completed. Please address the error\nabove and apply again to incrementally change your infrastructure." 
```

/test e2e-aws

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/724/pull-ci-openshift-installer-master-e2e-aws/1759/build-log.txt

File: pkg/tfvars/tfvars.go [hunk 2]
```diff
@@ -64,16 +63,9 @@ func TFVars(cfg *types.InstallConfig, bootstrapIgn, masterIgn string) ([]byte, e
 				}
 			}
 		case "worker":
-			config.Workers += replicas
 			if m.Platform.AWS != nil {
 				config.AWS.Worker = aws.Worker{
-					EC2Type:     m.Platform.AWS.InstanceType,
 					IAMRoleName: m.Platform.AWS.IAMRoleName,
-					WorkerRootVolume: aws.WorkerRootVolume{
```
Comment (smarterclayton) [inline pkg/tfvars/tfvars.go]: Is this why workers are now 16gb volumes?

Comment (wking) [reply pkg/tfvars/tfvars.go]: > Is this why workers are now 16gb volumes?

This is just passing data to Terraform, and worker creation is via the cluster-API now.  So I don't see a connection.

**Comment by smarterclayton** [comment]
/lgtm

To unblock all the queues

**Comment by smarterclayton** [comment]
45m of this run was tear down, and it looks like a big chunk of that was route53.  We have a ton of garbage in the api, probably needs to be cleaned

**Comment by wking** [comment]
>  We have a ton of garbage in the api, probably needs to be cleaned

I cleaned it ;)

**Comment by wking** [comment]
Hrm, this is going to be a problem with `t3.medium`:

```console
$ oc --namespace=openshift-cluster-api logs clusterapi-manager-controllers-559fc8d878-d5p65  --container=machine-controller
E1130 05:38:29.574505       1 actuator.go:96] error creating machine: error launching instance: error creating EC2 instance: InstanceLimitExceeded: You have requested more instances (21) than your current instance limit of 20 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.
        status code: 400, request id: 01da30c5-c310-48c0-9550-79cc78cccd7e
```

(from [this cluster][1]).  Can we get some serialization so we can limp through until we get that limit raised?  Switch to some other instance type where we have more power than `t2.medium` but higher limits than `t3.medium`?

[1]: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/730/pull-ci-openshift-installer-master-e2e-aws/1764

**Comment by wking** [comment]
#765 is in flight to get us on `m4.large` while we work on the `t3.medium` limits.

**Pull request merged**